### PR TITLE
Add Jose4j static init recording to OIDC

### LINKED ISSUE #41452: Quarkus OIDC makes first request very slow
### Describe the bug

Hi,

We use Quarkus OIDC for multi tenancy. We do not use Keycloak, we use our own authentication system. Therefore we set `quarkus.oidc.discovery-enabled` to false, which results in our OIDC config not getting loaded on startup, but on the first request. This makes the first request that gets done to our container (512 vCPU, 1024 memory) very slow (around 4 seconds).

Is there a way to make a healthcheck manually load the OIDC providers before the first request gets done to the container? Or could there be added a new configuration property to load the providers on startup (such as proposed in https://github.com/quarkusio/quarkus/issues/34890)

```
quarkus.oidc.enabled=true
quarkus.oidc.discovery-enabled=false
quarkus.oidc.use-blocking-dns-lookup=true
quarkus.oidc.bca.public-key=${OUR_PUBLIC_KEY}
quarkus.oidc.bca.token.audience=bsp
quarkus.oidc.bca.token.issued-at-required=false
quarkus.oidc.bca.application-type=service
quarkus.oidc.bca.token.signature-algorithm=rs256
quarkus.oidc.auth-server-url=${JWKS_LOCATION}
quarkus.oidc.jwks-path=${JWKS_LOCATION}
quarkus.oidc.token.issued-at-required=false
quarkus.oidc.token.signature-algorithm=es256
quarkus.oidc.application-type=service
```

We also have our own TenantResolverConfig, to get the tenant config based on the tenant:

```
@ApplicationScoped
public class TokenTenantResolver implements TenantConfigResolver {
  private static final String ISSUER_PATH = "iss";

  @ConfigProperty(name = "quarkus.oidc.bca.token.issuer")
  String bcaIssuer;

  @ConfigProperty(name = "quarkus.oidc.bca.public-key")
  String bcaPublicKey;

  @SneakyThrows
  public Uni<OidcTenantConfig> resolve(
      RoutingContext context, OidcRequestContext<OidcTenantConfig> requestContext) {
    final var jwtString = context.request().headers().get("Authorization");

    if (jwtString == null) {
      return Uni.createFrom().nullItem();
    }

    final var jwtHeaders = io.quarkus.oidc.runtime.OidcUtils.decodeJwtContent(jwtString);

    if (jwtHeaders != null && jwtHeaders.getValue(ISSUER_PATH).equals(bcaIssuer)) {
      OidcTenantConfig config = new OidcTenantConfig();
      config.setTenantId("bca");
      config.setPublicKey(bcaPublicKey);
      config.getToken().setIssuedAtRequired(false);
      config.setApplicationType(OidcTenantConfig.ApplicationType.SERVICE);
      return Uni.createFrom().item(config);
    } else {
      // resolve to default tenant config
      return Uni.createFrom().nullItem();
    }
  }
}
```

### Expected behavior

All the fields in our config for our OIDC tenants are "deterministic" as in, they could easily be loaded on startup. So initializing them on startup would be our expected behavior

### Actual behavior

The OIDC config does not get loaded on startup, resulting in intializing all OIDC classes etc on the first request, which makes the first request very slow.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

AWS ECS Fargate container

### Output of `java -version`

openjdk 21.0.1 2023-10-17 LTS OpenJDK Runtime Environment Corretto-21.0.1.12.1 (build 21.0.1+12-LTS) OpenJDK 64-Bit Server VM Corretto-21.0.1.12.1 (build 21.0.1+12-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.11.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

docker

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by sberyozkin**
- Fixes: #41452

### CODE REVIEW

**Comment by sberyozkin** [comment]
Hi @gsmet @aloubyansky, I've seen this Gradle build failure a few times before, it looked transient, I meant to open a separate issue

**Review by aloubyansky — APPROVED** [review]

File: extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/Jose4jRecorder.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.jose4j.jwa.AlgorithmFactoryFactory;
+
+import io.quarkus.runtime.annotations.Recorder;
+
+@Recorder
+public class Jose4jRecorder {
+
+    public void initialize() {
+        AlgorithmFactoryFactory.getInstance();
```
Comment (geoand) [inline extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/Jose4jRecorder.java:11]: This uses ` System.getProperty` which should never be used to configure behavior at native-image build time.

To be fair, it seems to use it only for logging, but it makes suspicious of what else this method does and it absolutely needs to be checked thoroughly

Comment (geoand) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/Jose4jRecorder.java:11]: > The rest of the method is sound, these properties are not having any impact.

If you have checked it thoroughly, that's fine with me.

Comment (sberyozkin) [reply extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/Jose4jRecorder.java:11]: @geoand Yeah, I searched if any of those properties can be seen somewhere else in that code :-). The rest is just loading supported algorithms, none of them would need a version/vendor/javahome property. I suppose, purely for debugging purposes, seeing Java version and vendor can be useful in context of the Jose4 initialization as a support for various algorithms may vary...
Agreed it was not really essential though

**Comment by sberyozkin** [comment]
Let me merge now that we've discussed Georgios's comment

**Pull request merged**